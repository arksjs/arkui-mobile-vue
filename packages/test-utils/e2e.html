<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
