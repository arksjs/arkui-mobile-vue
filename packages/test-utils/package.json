{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
