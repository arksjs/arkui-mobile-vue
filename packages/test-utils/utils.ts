export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
