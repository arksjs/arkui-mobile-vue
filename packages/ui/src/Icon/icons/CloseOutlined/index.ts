export { default as default } from './CloseOutlined.vue'
