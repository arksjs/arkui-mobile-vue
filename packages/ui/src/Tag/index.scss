@import '../style/var.scss';
@import '../style/mixin.scss';

.ta-tag {
  --ta-color: transparent;
  --ta-black-color: transparent;
  --ta-white-color: #{$white-color};
  --ta-front-color: var(--ta-white-color);
  --ta-tag-icon-size: 16px;

  display: inline-flex;
  align-items: center;
  height: 24px;
  padding: 0 12px;
  border-radius: 2px;
  font-size: 12px;
  color: var(--ta-front-color);
  background-color: var(--ta-color);
  border-style: solid;
  border-width: 0;
  border-color: transparent;
  box-sizing: border-box;
  user-select: none;

  + .ta-tag {
    margin-left: 12px;
  }

  &_close {
    cursor: pointer;
    margin: 0 -2px 0 2px;
    appearance: none;
    padding: 0;
    border: 0;
    background: none;
    width: 16px;
    height: 16px;
    display: block;

    .ta-icon {
      display: block;

      --ta-icon-size: var(--ta-tag-icon-size);
      --ta-icon-color: var(--ta-front-color);
    }
  }

  &.pattern--light {
    --ta-front-color: var(--ta-black-color);
  }

  &.pattern--plain {
    background-color: var(--ta-white-color);
    border-color: var(--ta-color);

    --ta-front-color: var(--ta-color);
  }

  &.size--large {
    .ta-tag_close {
      margin: 0 -4px 0 4px;
    }

    &.pattern--plain {
      padding: 0 11px;
      border-width: 1px;
    }
  }

  &.size--middle {
    height: 18px;
    font-size: 10px;
    padding: 0 8px;

    --ta-tag-icon-size: 12px;

    &.pattern--plain {
      padding: 0 7px;
      border-width: 1px;
    }
  }

  &.size--small {
    height: 14px;
    font-size: 9px;
    padding: 0 4px;

    --ta-tag-icon-size: 10px;

    &.pattern--plain {
      padding: 0 3px;
      border-width: 1px;
    }
  }

  &.type--primary {
    --ta-color: #{$primary-color};
    --ta-black-color: #{$primary-color};

    &.pattern--dark {
      --ta-front-color: #fff;
    }

    &.pattern--light {
      --ta-color: var(--ta-color-primary-lighter);
    }
  }

  &.type--success {
    --ta-color: #{$success-color};
    --ta-black-color: #{$success-color};

    &.pattern--dark {
      --ta-front-color: #fff;
    }

    &.pattern--light {
      --ta-color: var(--ta-color-success-lighter);
    }
  }

  &.type--warning {
    --ta-color: #{$warning-color};
    --ta-black-color: #{$warning-color};

    &.pattern--dark {
      --ta-front-color: #fff;
    }

    &.pattern--light {
      --ta-color: var(--ta-color-warning-lighter);
    }
  }

  &.type--danger {
    --ta-color: #{$danger-color};
    --ta-black-color: #{$danger-color};

    &.pattern--dark {
      --ta-front-color: #fff;
    }

    &.pattern--light {
      --ta-color: var(--ta-color-danger-lighter);
    }
  }

  &.type--default {
    --ta-color: #{$font-color};
    --ta-black-color: #{$font-color};

    &.pattern--light {
      --ta-color: #{$background-color};
    }

    &.pattern--plain {
      border-color: $border-color;
    }
  }

  &.disabled {
    cursor: not-allowed;
    opacity: 0.2;

    .ta-icon {
      cursor: not-allowed;
    }
  }

  &:not(.disabled) {
    .ta-tag_close:active {
      border-radius: 2px;
      @include mask-background(0.16);
    }
  }
}
