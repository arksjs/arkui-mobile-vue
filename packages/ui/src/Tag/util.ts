import {
  getColorObject,
  SIZE_TYPES,
  STATE_TYPES,
  getEnumsValue,
  type CSSProperties
} from '../helpers'
import type { TagProps, PatternType } from './types'

export const TAG_PATTERN_TYPES: PatternType[] = ['light', 'dark', 'plain']

export const getClasses = (props: TagProps) => {
  const { hasColor, isDark } = getColorObject(props.color)

  return [
    'ta-tag',
    'type--' + (hasColor ? STATE_TYPES[1] : getEnumsValue(STATE_TYPES, props.type)),
    'size--' + getEnumsValue(SIZE_TYPES, props.size),
    'pattern--' +
      (hasColor && props.pattern !== 'plain'
        ? isDark
          ? 'dark'
          : 'light'
        : getEnumsValue(TAG_PATTERN_TYPES, props.pattern)),
    { disabled: !!props.disabled }
  ]
}

export const getStyles = (color?: string) => {
  const styles: CSSProperties = {}

  const colorObj = getColorObject(color)

  if (colorObj.hasColor) {
    styles[`--ta-color`] = colorObj.varColor
    styles[`--ta-black-color`] = colorObj.varBlackColor
  }

  return styles
}
