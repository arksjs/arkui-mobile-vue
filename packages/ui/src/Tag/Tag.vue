<template>
  <div :class="classes" :style="styles" ref="root">
    <slot></slot>
    <button class="ta-tag_close" v-if="closable">
      <Icon
        :icon="CloseOutlined"
        @mousedown.stop="noop"
        @touchstart.stop="noop"
        @click="onClose"
      />
    </button>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, shallowRef, type PropType } from 'vue'
import { Icon } from '../Icon'
import {
  colorValidator,
  createEnumsValidator,
  SIZE_TYPES,
  STATE_TYPES,
  kebabCase2CamelCase,
  noop,
  returnTrue,
  type PropsToEmits,
  type SizeType,
  type StateType
} from '../helpers'
import { useLongPress } from '../hooks'
import type { PatternType, TagEmits } from './types'
import CloseOutlined from '../Icon/icons/CloseOutlined'
import { getStyles, getClasses, TAG_PATTERN_TYPES } from './util'

export default defineComponent({
  name: 'ta-tag',
  components: { Icon },
  props: {
    size: {
      type: String as PropType<SizeType>,
      validator: createEnumsValidator(SIZE_TYPES)
    },
    type: {
      type: String as PropType<StateType>,
      validator: createEnumsValidator(STATE_TYPES)
    },
    // 款式
    pattern: {
      type: String as PropType<PatternType>,
      validator: createEnumsValidator(TAG_PATTERN_TYPES)
    },
    // 可关闭的
    closable: {
      type: Boolean,
      default: false
    },
    // 禁用的
    disabled: {
      type: Boolean,
      default: false
    },
    color: {
      type: String,
      validator: colorValidator
    }
  },
  emits: {
    close: returnTrue,
    click: returnTrue,
    longPress: returnTrue
  } as PropsToEmits<TagEmits>,
  setup(props, { emit }) {
    const root = shallowRef<HTMLElement | null>(null)

    function onClose() {
      if (!props.disabled) {
        emit('close')
      }
    }

    const classes = computed(() => getClasses(props))
    const styles = computed(() => getStyles(props.color))

    useLongPress(root, e => {
      if (!props.disabled) {
        emit(kebabCase2CamelCase(e.type) as 'longPress')
      }
    })

    return {
      root,
      classes,
      styles,
      noop,
      onClose,
      CloseOutlined
    }
  }
})
</script>
