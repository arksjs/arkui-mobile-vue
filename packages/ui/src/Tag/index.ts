import { withInstall } from '../helpers/with-install'
import Tag from './Tag.vue'

export { Tag }
export const install = withInstall(Tag)
export default Tag
