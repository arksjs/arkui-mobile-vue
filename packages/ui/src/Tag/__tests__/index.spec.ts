import { describe, test, expect } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'
import { Tag } from '@/Tag'
import CloseOutlined from '@/Icon/icons/CloseOutlined'

describe('Tag', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(Tag, {})
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('pattern', async () => {
    const wrapper = mount(Tag, {})
    expect(wrapper.classes()).toContain('pattern--light')

    await wrapper.setProps({
      pattern: 'light'
    })
    expect(wrapper.classes()).toContain('pattern--light')

    await wrapper.setProps({
      pattern: 'dark'
    })
    expect(wrapper.classes()).toContain('pattern--dark')

    await wrapper.setProps({
      pattern: 'plain'
    })
    expect(wrapper.classes()).toContain('pattern--plain')
  })

  test('size', async () => {
    const wrapper = mount(Tag, {})
    expect(wrapper.classes()).toContain('size--middle')

    await wrapper.setProps({
      size: 'small'
    })
    expect(wrapper.classes()).toContain('size--small')

    await wrapper.setProps({
      size: 'middle'
    })
    expect(wrapper.classes()).toContain('size--middle')

    await wrapper.setProps({
      size: 'large'
    })
    expect(wrapper.classes()).toContain('size--large')
  })

  test('color', async () => {
    const wrapper = mount(Tag, {
      props: {
        color: '#6667AB'
      }
    })

    expect(wrapper.attributes('style')).toContain('--ta-color: #6667AB;')
    expect(wrapper.attributes('style')).toContain('--ta-black-color: #090812;')

    await wrapper.setProps({
      color: '#E2C0BF'
    })

    expect(wrapper.attributes('style')).toContain('--ta-color: #E2C0BF;')
    expect(wrapper.attributes('style')).toContain('--ta-black-color: #493234;')
  })

  test('closable', () => {
    const wrapper = mount(Tag, {
      props: {
        closable: true
      }
    })
    expect(wrapper.findComponent(CloseOutlined).exists()).toBeTruthy()
  })
})
