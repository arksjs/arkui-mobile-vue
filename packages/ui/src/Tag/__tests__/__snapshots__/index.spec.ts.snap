// Vitest Snapshot v1

exports[`Tag > snapshot 1`] = `
"<div class=\\"ta-tag type--default size--middle pattern--light\\">
  <!--v-if-->
</div>"
`;
